data_structures/src/staking/stakes.rs

```diff
@@ -294,17 +299,50 @@ where
             .get_mut(&validator)
             .ok_or(StakesError::ValidatorNotFound { validator })?;
         stakes.iter_mut().for_each(|stake| {
+            let old_epoch = stake.value.write().unwrap().epochs.get(capability);
+            let update_epoch = (current_epoch - old_epoch) / Epoch::from(reset_factor);
             stake
                 .value
                 .write()
                 .unwrap()
                 .epochs
-                .update(capability, current_epoch)
+                .update(capability, old_epoch + update_epoch);
         });
 
         Ok(())
     }
 
+    /// Add a reward to the validator's balance
+    pub fn add_reward<ISK>(
+        &mut self,
+        validator: ISK,
+        coins: Coins,
+        current_epoch: Epoch,
+    ) -> StakesResult<Coins, Address, Coins, Epoch>
+    where
+        ISK: Into<Address>,
+    {
+        let validator = validator.into();
+
+        let stakes = self
+            .by_validator
+            .get_mut(&validator)
+            .ok_or(StakesError::ValidatorNotFound { validator })?;
+
+        let distribute_coins = coins / Coins::from(stakes.len() as u64);
+        let remainder_coins = coins % Coins::from(stakes.len() as u64);
+
+        stakes.iter_mut().for_each(|stake| {
+            let _ = stake.value.write().unwrap().add_stake(
+                distribute_coins,
+                current_epoch,
+                Some(0.into()),
+            );
+        });
+
+        Ok(remainder_coins)
+    }
```

Comment:
This essentially implements a model where validators accrue rewards that are bound to become withdrawable by multiple addresses in proportion to the amounts staked in connection to those different withdrawer addresses.

I'm trying to go for a different model where stake entries are operated indepedently, i.e. a validators must compute eligibility separately for each of the stake entries in which they're designated as the validator. Then, a reward obtained by leveraging the power of one entry goes only to that one entry. Any other stake entries operated by the same validator should remain unaffected both in terms of rewards and penalties.

This does not only simplify rewards and withdrawals, but also paves the way for a future 2.X upgrade in which we open the door to allowing an operator to operate stake entries for multiple withdrawers, by means of simply removing the "1 operator = 1 withdrawer" constraint introduced here.

Why do I favor that approach? Because (with enough support from the community) we can pull off the trick of adding basic support for delegation in record time after 2.0 release, and with near-zero development effort.

Reply:
> This essentially implements a model where validators accrue rewards that are bound to become withdrawable by multiple addresses in proportion to the amounts staked in connection to those different withdrawer addresses.

Correct, that was the goal.

> I'm trying to go for a different model where stake entries are operated indepedently, i.e. a validators must compute eligibility separately for each of the stake entries in which they're designated as the validator. Then, a reward obtained by leveraging the power of one entry goes only to that one entry. Any other stake entries operated by the same validator should remain unaffected both in terms of rewards and penalties.

I disagree with the efficacy of this model for two reasons:
1) It complicates the power and rank functions. Rather than a simple sum of powers to calculate eligibility for a single validator, they now have to be split it up.
2) In my opinion, this defeats the main purpose of delegation: small stakers combining their power to operate as if they were a high-value validator. If each staker's power and eligibility are only proportional to their own stake, they don't really gain a whole lot by staking on a validator they do not control. Time to a block will stay as large as if they were validating with only their own stake on a self-owned validator and rather than a somewhat continuous stream of income (because their combined power significantly increases their eligibility and they split and share rewards), they need to wait a long time to get a single higher-value reward. In other words, this keeps pretty much the same variability issue we have in the current mining model.

> This does not only simplify rewards and withdrawals, but also paves the way for a future 2.X upgrade in which we open the door to allowing an operator to operate stake entries for multiple withdrawers, by means of simply removing the "1 operator = 1 withdrawer" constraint introduced here.

I don't see why this model is better for enabling a multi-staker, single-validator model. I think most of the infrastructure (such as a vector of stake entries in the `by_validator` tracker structure) is already in place.

Reply:
> 1. It complicates the power and rank functions. Rather than a simple sum of powers to calculate eligibility for a single validator, they now have to be split it up.

It does not complicate such functions at all. Each stake entry in `by_key` has its own power and rank. Each entry has its own eligibility, and they are operated independently. There's no need for adding up stakes or anything like that because under that model there's no such thing as "a validator's eligibillity" but rather "a stake entry's eligibility". The only downside is that each validator must compute separate VRFs for each of the withdrawers.

> 2. In my opinion, this defeats the main purpose of delegation: small stakers combining their power to operate as if they were a high-value validator. If each staker's power and eligibility are only proportional to their own stake, they don't really gain a whole lot by staking on a validator they do not control. Time to a block will stay as large as if they were validating with only their own stake on a self-owned validator and rather than a somewhat continuous stream of income (because their combined power significantly increases their eligibility and they split and share rewards), they need to wait a long time to get a single higher-value reward. In other words, this keeps pretty much the same variability issue we have in the current mining model.

I'm assuming that the kind of PoS that we're implementing is fairly linear, so as you have described, the main point of pooling stake is to increase the frequency of rewards (even if long term the total amount of rewards is the same).

I push for this "independent stake entries" model because I can see a wide demographic of people in our community who will probably pass on running their own validator, but would be willing to delegate their coins into other community members' validators, knowing that they can't really steal their coins. At the same time, as I hinted before, I'm not pushing for supporting delegation just right now, but in a future update, once the wit/2 network is fully bootstrapped.

> I don't see why this model is better for enabling a multi-staker, single-validator model. I think most of the infrastructure (such as a vector of stake entries in the `by_validator` tracker structure) is already in place.

I see a clear advantage when it comes to rewards and withdrawals. Not having to split rewards and handle remainders like above is pretty convenient.

Reply:
> It does not complicate such functions at all. Each stake entry in by_key has its own power and rank. Each entry has its own eligibility, and they are operated independently. There's no need for adding up stakes or anything like that because under that model there's no such thing as "a validator's eligibillity" but rather "a stake entry's eligibility". The only downside is that each validator must compute separate VRFs for each of the withdrawers.

Yes, you are right. I was still reasoning from a `by_validator` iterator, but you could also iterate on `by_key` which means you do not have to flatten vectors etc.

> I'm assuming that the kind of PoS that we're implementing is fairly linear, so as you have described, the main point of pooling stake is to increase the frequency of rewards (even if long term the total amount of rewards is the same).

But if you are handling each stake entry as a different entity, the frequency of rewards does not increase? That's my whole point. Someone staking 100k WIT on a 10M WIT validator will not see a more continuous stream of rewards due to that validator splitting block rewards, but rather a single "high" block rewards after a long time.

> I push for this "independent stake entries" model because I can see a wide demographic of people in our community who will probably pass on running their own validator, but would be willing to delegate their coins into other community members' validators, knowing that they can't really steal their coins.

I agree with that point, that will likely happen.

> I see a clear advantage when it comes to rewards and withdrawals. Not having to split rewards and handle remainders like above is pretty convenient.

I see how it simplifies the reward function, I do not see how it simplifies the withdrawals. I have no idea what would have to change there because that function should not care about unified or split rewards in another step.

Reply:
Resolved by simplifying the `add_reward` function until a later release as discussed offline.